File: euler/py/src/e004.py
#!/usr/bin/env python
# by protago90

from src.utils import promptify


def run_e4v1(x: int) -> int:
    rec = 1
    t, tt  = 10**(x-1), 10**x - 1
    for n in range(t, tt):
        for m in range(n, tt):
            p = n * m
            if p == int(str(p)[::-1]) & p > rec:
                rec = p 
    return rec 


@promptify
def solve_e4() -> int:
    '''
    Largest polindrome product  
    A palindromic number reads the same both ways. The largest palindrome made from
    the product of two 2-digit numbers is 9009 = 91 × 99.
    > Find the largest palindrome made from the product of two 3-digit numbers.
    ''' 
    return run_e4v1(3)


if __name__ == '__main__':
    solve_e4()
    # >> the anwser for the #4 euler problem is >906609<; computed in 0.2008s ∎

    # from src.utils import timeitfy
    # timeitfy([run_e4v1], args=[3], i=10000)

File: euler/c/src/e004.c
// by protago90

#include "utils.h"
#include <math.h>


int run_e4v1(int x) {
    int rec = 1;
    int p_rev, p;
    int lim = pow(10, x) - 1;
    int n   = pow(10, x-1);
    int m;
    for (n; n < lim; n++) {
        for (m = n; m < lim; m++) {
            int tmp = n*m;
            p = tmp;
            p_rev = 0;
            while (tmp != 0) {
                p_rev = p_rev*10 + tmp % 10;
                tmp /= 10;
            }
            if (p == p_rev && p > rec) { rec = p; }
        }
    } 	
    return rec;
}


long solve_e4() {
   /***
    * Largest polindrome product
    * A palindromic number reads the same both ways. The largest palindrome made from
    * the product of two 2-digit numbers is 9009 = 91 × 99.
    * > Find the largest palindrome made from the product of two 3-digit numbers.
    **/
    return run_e4v1(3); 
}


// int main() {
//     promptify(solve_e4, 4);
//  // >> the anwser for the #4 euler problem is >906609<; computed in 0.0111s ∎
//     return 0;
// }